<html>
<head>
  <title>VEP summary</title>
  <script type="text/javascript" src="http://www.google.com/jsapi"></script>
  <script type="text/javascript">
    google.load('visualization', '1', {packages: ['corechart','table']});
  </script>
  <script type="text/javascript">
    
    function init() {
      // charts
      var var_class_pie = drawPie('var_class_pie', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['indel',120],['sequence_alteration',2084],['insertion',27199],['deletion',32510],['SNV',314773]]), null);
var var_class_table = drawTable('var_class_table', 'Variant classes', google.visualization.arrayToDataTable([['Variant class','Count'],['indel',120],['sequence_alteration',2084],['insertion',27199],['deletion',32510],['SNV',314773]]));

          google.visualization.events.addListener(var_class_pie, 'select', function() {
            var_class_table.setSelection(var_class_pie.getSelection());
          });
          google.visualization.events.addListener(var_class_table, 'select', function() {
            var_class_pie.setSelection(var_class_table.getSelection());
          });
        var var_cons_pie = drawPie('var_cons_pie', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',120],['splice_donor_variant',122],['stop_gained',147],['frameshift_variant',327],['stop_lost',47],['start_lost',33],['inframe_insertion',168],['inframe_deletion',191],['protein_altering_variant',2],['missense_variant',12391],['splice_region_variant',3300],['synonymous_variant',11885],['stop_retained_variant',18],['coding_sequence_variant',5],['mature_miRNA_variant',8],['5_prime_UTR_variant',3811],['3_prime_UTR_variant',10705],['non_coding_transcript_exon_variant',24460],['intron_variant',215819],['upstream_gene_variant',12725],['downstream_gene_variant',9381],['intergenic_variant',71021]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {}, {color: "#ffd700"}, {color: "#ff7f50"}, {color: "#76ee00"}, {color: "#76ee00"}, {color: "#458b00"}, {}, {}, {}, {}, {color: "#02599c"}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {color: "gray"}]});
var var_cons_table = drawTable('var_cons_table', 'Consequences (most severe)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_acceptor_variant',120],['splice_donor_variant',122],['stop_gained',147],['frameshift_variant',327],['stop_lost',47],['start_lost',33],['inframe_insertion',168],['inframe_deletion',191],['protein_altering_variant',2],['missense_variant',12391],['splice_region_variant',3300],['synonymous_variant',11885],['stop_retained_variant',18],['coding_sequence_variant',5],['mature_miRNA_variant',8],['5_prime_UTR_variant',3811],['3_prime_UTR_variant',10705],['non_coding_transcript_exon_variant',24460],['intron_variant',215819],['upstream_gene_variant',12725],['downstream_gene_variant',9381],['intergenic_variant',71021]]));

          google.visualization.events.addListener(var_cons_pie, 'select', function() {
            var_cons_table.setSelection(var_cons_pie.getSelection());
          });
          google.visualization.events.addListener(var_cons_table, 'select', function() {
            var_cons_pie.setSelection(var_cons_table.getSelection());
          });
        var consequences_pie = drawPie('consequences_pie', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',293],['splice_acceptor_variant',266],['stop_gained',269],['frameshift_variant',705],['stop_lost',67],['start_lost',88],['inframe_insertion',437],['inframe_deletion',591],['missense_variant',32457],['protein_altering_variant',4],['splice_region_variant',15701],['incomplete_terminal_codon_variant',2],['synonymous_variant',38915],['stop_retained_variant',47],['start_retained_variant',14],['coding_sequence_variant',67],['mature_miRNA_variant',10],['5_prime_UTR_variant',10225],['3_prime_UTR_variant',33226],['non_coding_transcript_exon_variant',57865],['intron_variant',1325864],['NMD_transcript_variant',172428],['non_coding_transcript_variant',387419],['upstream_gene_variant',247636],['downstream_gene_variant',291536],['intergenic_variant',71296]]), {slices: [{color: "#ff7f50"}, {color: "#ff7f50"}, {color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {color: "#ffd700"}, {}, {color: "#ff7f50"}, {color: "#ff00ff"}, {color: "#76ee00"}, {color: "#76ee00"}, {}, {color: "#458b00"}, {}, {}, {}, {}, {color: "#02599c"}, {}, {}, {color: "#a2b5cd"}, {color: "#a2b5cd"}, {color: "gray"}]});
var consequences_table = drawTable('consequences_table', 'Consequences (all)', google.visualization.arrayToDataTable([['Consequence type','Count'],['splice_donor_variant',293],['splice_acceptor_variant',266],['stop_gained',269],['frameshift_variant',705],['stop_lost',67],['start_lost',88],['inframe_insertion',437],['inframe_deletion',591],['missense_variant',32457],['protein_altering_variant',4],['splice_region_variant',15701],['incomplete_terminal_codon_variant',2],['synonymous_variant',38915],['stop_retained_variant',47],['start_retained_variant',14],['coding_sequence_variant',67],['mature_miRNA_variant',10],['5_prime_UTR_variant',10225],['3_prime_UTR_variant',33226],['non_coding_transcript_exon_variant',57865],['intron_variant',1325864],['NMD_transcript_variant',172428],['non_coding_transcript_variant',387419],['upstream_gene_variant',247636],['downstream_gene_variant',291536],['intergenic_variant',71296]]));

          google.visualization.events.addListener(consequences_pie, 'select', function() {
            consequences_table.setSelection(consequences_pie.getSelection());
          });
          google.visualization.events.addListener(consequences_table, 'select', function() {
            consequences_pie.setSelection(consequences_table.getSelection());
          });
        var coding_pie = drawPie('coding_pie', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',269],['frameshift_variant',705],['stop_lost',67],['start_lost',88],['inframe_insertion',437],['inframe_deletion',591],['protein_altering_variant',4],['missense_variant',32457],['incomplete_terminal_codon_variant',2],['synonymous_variant',38915],['start_retained_variant',14],['stop_retained_variant',47],['coding_sequence_variant',67]]), {slices: [{color: "#ff0000"}, {color: "#ff69b4"}, {color: "#ff0000"}, {}, {color: "#ff69b4"}, {color: "#ff69b4"}, {}, {color: "#ffd700"}, {color: "#ff00ff"}, {color: "#76ee00"}, {}, {color: "#76ee00"}, {color: "#458b00"}]});
var coding_table = drawTable('coding_table', 'Coding consequences', google.visualization.arrayToDataTable([['Consequence type','Count'],['stop_gained',269],['frameshift_variant',705],['stop_lost',67],['start_lost',88],['inframe_insertion',437],['inframe_deletion',591],['protein_altering_variant',4],['missense_variant',32457],['incomplete_terminal_codon_variant',2],['synonymous_variant',38915],['start_retained_variant',14],['stop_retained_variant',47],['coding_sequence_variant',67]]));

          google.visualization.events.addListener(coding_pie, 'select', function() {
            coding_table.setSelection(coding_pie.getSelection());
          });
          google.visualization.events.addListener(coding_table, 'select', function() {
            coding_pie.setSelection(coding_table.getSelection());
          });
        var chr_bar = drawBar('chr_bar', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['chr1',32539],['chr10',18076],['chr11',18868],['chr11_KI270721v1_random',3],['chr12',18384],['chr13',10944],['chr14',11031],['chr14_GL000009v2_random',58],['chr14_GL000194v1_random',122],['chr14_GL000225v1_random',2364],['chr14_KI270722v1_random',45],['chr14_KI270723v1_random',135],['chr14_KI270724v1_random',150],['chr14_KI270725v1_random',16],['chr15',12404],['chr15_KI270727v1_random',46],['chr16',13344],['chr16_KI270728v1_random',172],['chr17',15283],['chr17_GL000205v2_random',134],['chr17_KI270729v1_random',139],['chr17_KI270730v1_random',53],['chr18',7901],['chr19',15060],['chr1_KI270706v1_random',24],['chr1_KI270707v1_random',4],['chr1_KI270708v1_random',11],['chr1_KI270709v1_random',75],['chr1_KI270710v1_random',8],['chr1_KI270711v1_random',65],['chr1_KI270712v1_random',13],['chr1_KI270713v1_random',67],['chr1_KI270714v1_random',28],['chr2',28283],['chr20',9556],['chr21',5977],['chr22',7636],['chr22_KI270731v1_random',12],['chr22_KI270732v1_random',32],['chr22_KI270733v1_random',142],['chr22_KI270734v1_random',11],['chr22_KI270735v1_random',27],['chr22_KI270736v1_random',55],['chr22_KI270737v1_random',4],['chr22_KI270738v1_random',34],['chr22_KI270739v1_random',17],['chr2_KI270716v1_random',3],['chr2_KI270775v1_alt',1],['chr3',22128],['chr3_GL000221v1_random',20],['chr4',20043],['chr4_GL000008v2_random',65],['chr5',18221],['chr5_GL000208v1_random',20],['chr6',21641],['chr7',19911],['chr8',14891],['chr9',15681],['chr9_KI270718v1_random',6],['chr9_KI270719v1_random',41],['chr9_KI270720v1_random',43],['chrM',28],['chrUn_GL000195v1',62],['chrUn_GL000213v1',8],['chrUn_GL000214v1',39],['chrUn_GL000216v2',574],['chrUn_GL000218v1',116],['chrUn_GL000219v1',208],['chrUn_GL000220v1',16],['chrUn_GL000224v1',75],['chrUn_KI270330v1',8],['chrUn_KI270333v1',17],['chrUn_KI270336v1',12],['chrUn_KI270337v1',16],['chrUn_KI270438v1',375],['chrUn_KI270442v1',515],['chrUn_KI270466v1',24],['chrUn_KI270519v1',127],['chrUn_KI270538v1',51],['chrUn_KI270742v1',63],['chrUn_KI270743v1',25],['chrUn_KI270744v1',129],['chrUn_KI270745v1',2],['chrUn_KI270747v1',5],['chrUn_KI270749v1',10],['chrUn_KI270750v1',27],['chrUn_KI270751v1',48],['chrUn_KI270753v1',2],['chrUn_KI270754v1',23],['chrUn_KI270756v1',12],['chrUn_KI270757v1',85],['chrX',10558],['chrY',1359]]), {legend: {position: "none"}});
var chr_table = drawTable('chr_table', 'Variants by chromosome', google.visualization.arrayToDataTable([['Chromosome','Count'],['chr1',32539],['chr10',18076],['chr11',18868],['chr11_KI270721v1_random',3],['chr12',18384],['chr13',10944],['chr14',11031],['chr14_GL000009v2_random',58],['chr14_GL000194v1_random',122],['chr14_GL000225v1_random',2364],['chr14_KI270722v1_random',45],['chr14_KI270723v1_random',135],['chr14_KI270724v1_random',150],['chr14_KI270725v1_random',16],['chr15',12404],['chr15_KI270727v1_random',46],['chr16',13344],['chr16_KI270728v1_random',172],['chr17',15283],['chr17_GL000205v2_random',134],['chr17_KI270729v1_random',139],['chr17_KI270730v1_random',53],['chr18',7901],['chr19',15060],['chr1_KI270706v1_random',24],['chr1_KI270707v1_random',4],['chr1_KI270708v1_random',11],['chr1_KI270709v1_random',75],['chr1_KI270710v1_random',8],['chr1_KI270711v1_random',65],['chr1_KI270712v1_random',13],['chr1_KI270713v1_random',67],['chr1_KI270714v1_random',28],['chr2',28283],['chr20',9556],['chr21',5977],['chr22',7636],['chr22_KI270731v1_random',12],['chr22_KI270732v1_random',32],['chr22_KI270733v1_random',142],['chr22_KI270734v1_random',11],['chr22_KI270735v1_random',27],['chr22_KI270736v1_random',55],['chr22_KI270737v1_random',4],['chr22_KI270738v1_random',34],['chr22_KI270739v1_random',17],['chr2_KI270716v1_random',3],['chr2_KI270775v1_alt',1],['chr3',22128],['chr3_GL000221v1_random',20],['chr4',20043],['chr4_GL000008v2_random',65],['chr5',18221],['chr5_GL000208v1_random',20],['chr6',21641],['chr7',19911],['chr8',14891],['chr9',15681],['chr9_KI270718v1_random',6],['chr9_KI270719v1_random',41],['chr9_KI270720v1_random',43],['chrM',28],['chrUn_GL000195v1',62],['chrUn_GL000213v1',8],['chrUn_GL000214v1',39],['chrUn_GL000216v2',574],['chrUn_GL000218v1',116],['chrUn_GL000219v1',208],['chrUn_GL000220v1',16],['chrUn_GL000224v1',75],['chrUn_KI270330v1',8],['chrUn_KI270333v1',17],['chrUn_KI270336v1',12],['chrUn_KI270337v1',16],['chrUn_KI270438v1',375],['chrUn_KI270442v1',515],['chrUn_KI270466v1',24],['chrUn_KI270519v1',127],['chrUn_KI270538v1',51],['chrUn_KI270742v1',63],['chrUn_KI270743v1',25],['chrUn_KI270744v1',129],['chrUn_KI270745v1',2],['chrUn_KI270747v1',5],['chrUn_KI270749v1',10],['chrUn_KI270750v1',27],['chrUn_KI270751v1',48],['chrUn_KI270753v1',2],['chrUn_KI270754v1',23],['chrUn_KI270756v1',12],['chrUn_KI270757v1',85],['chrX',10558],['chrY',1359]]));

          google.visualization.events.addListener(chr_bar, 'select', function() {
            chr_table.setSelection(chr_bar.getSelection());
          });
          google.visualization.events.addListener(chr_table, 'select', function() {
            chr_bar.setSelection(chr_table.getSelection());
          });
        var protein_bar = drawBar('protein_bar', 'Position in protein', google.visualization.arrayToDataTable([['Position in protein (percentile)','Count'],['00-10%',7459],['10-20%',7531],['20-30%',7745],['30-40%',7147],['40-50%',7320],['50-60%',7216],['60-70%',7041],['70-80%',7283],['80-90%',7083],['90-100%',7742]]), {hAxis: {title: "Position in protein (percentile)", textStyle: {fontSize: 10}}, legend: {position: "none"}});

    }
    
    function drawPie(id, title, data, options) {    
      var pie = new google.visualization.PieChart(document.getElementById(id));
      pie.draw(data, options);
      return pie;
    }
    function drawBar(id, title, data, options) {
      var bar = new google.visualization.ColumnChart(document.getElementById(id));
      bar.draw(data, options);
      return bar;
    }
    function drawTable(id, title, data) {
      var table = new google.visualization.Table(document.getElementById(id));
      table.draw(data, null);
      return table;
    }
    function drawLine(id, title, data, options) {
      var line = new google.visualization.LineChart(document.getElementById(id));
      line.draw(data, options);
      return line;
    }
    function drawArea(id, title, data, options) {
      var area = new google.visualization.AreaChart(document.getElementById(id));
      area.draw(data, options);
      return area;
    }
    google.setOnLoadCallback(init);
  </script>
  
  
  <style type="text/css">
    body {
      font-family: arial, sans-serif;
      margin: 0px;
      padding: 0px;
    }
    
    a {color: #36b;}
    a.visited {color: #006;}
    
    .stats_table {
      margin: 5px;
    }
    
    td {
      padding: 5px;
    }

    th.gradient {
      height: auto;
    }

    .stats_table tr:nth-child(odd) {
      background-color: #f0f0f0;
    }
    
    h3 {
      color: #666;
    }
    
    .masthead {
      background-color: white;
      color: rgb(204, 221, 255);
      height: 80px;
      width: 100%;
      padding: 0px;
    }
    
    .main {
      padding: 10px;
    }
    
    .gradient {
      background: #333366; /* Old browsers */
      background: -moz-linear-gradient(left,  #333366 0%, #ffffff 100%); /* FF3.6+ */
      background: -webkit-gradient(linear, left top, right top, color-stop(0%,#333366), color-stop(100%,#ffffff)); /* Chrome,Safari4+ */
      background: -webkit-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Chrome10+,Safari5.1+ */
      background: -o-linear-gradient(left,  #333366 0%,#ffffff 100%); /* Opera 11.10+ */
      background: -ms-linear-gradient(left,  #333366 0%,#ffffff 100%); /* IE10+ */
      background: linear-gradient(to right,  #333366 0%,#ffffff 100%); /* W3C */
      filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#333366', endColorstr='#ffffff',GradientType=1 ); /* IE6-9 */
      
      padding: 0px;
      height: 80px;
      width: 700px;
    }
    
    .main_content {
      margin-left: 300px;
    }
    
    .sidemenu {
      width: 260px;
      position: fixed;
      border-style: solid;
      border-width: 2px;
      border-color: rgb(51, 51, 102);
    }
    
    .sidemenu_head {
      width: 250px;
      background-color: rgb(51, 51, 102);
      color: rgb(204, 221, 255);
      padding: 5px;
    }
    
    .sidemenu_body {
      width: 250px;
      padding: 5px;
    }
  </style>
</head>
<body>
<div id="masthead" class="masthead">
  <div style="float: left; display: inline; padding: 10px; height: 80px;">
    <a href="http://www.ensembl.org/"><img src="http://static.ensembl.org/i/e-ensembl.png" width="185" height="45" alt="Ensembl"></a>
  </div>
  
  <div style="float: right; display: inline; height: 80px; background: white; padding: 10px;">
    <a href="http://www.ensembl.org/vep"><img src="http://www.ensembl.org/img/vep_logo.png" width="111" height="50" alt="VEP"></a>
  </div>
  <div class="gradient">
  </div>
</div>
<div class="main">
<div class="sidemenu"><div class="sidemenu_head">Links</div><div class="sidemenu_body"><ul><li><a href="#masthead">Top of page</a></li><li><a href="#run_stats">VEP run statistics</a></li><li><a href="#gen_stats">General statistics</a></li><li><a href="#var_class">Variant classes</a></li><li><a href="#var_cons">Consequences (most severe)</a></li><li><a href="#consequences">Consequences (all)</a></li><li><a href="#coding">Coding consequences</a></li><li><a href="#chr">Variants by chromosome</a></li><li><a href="#protein">Position in protein</a></li></ul></div></div><div class='main_content'><h3 id="run_stats">VEP run statistics</h3><table class="stats_table"><tr><td>VEP version (API)</td><td> 102 (102)</td></tr><tr><td>Annotation sources</td><td>Cache: /opt/vep/.vep/homo_sapiens/102_GRCh38</td></tr><tr><td>Species</td><td>homo_sapiens</td></tr><tr><td>Command line options</td><td><pre>--input_file /opt/vep/.vep/NSAIDS_0011.1d_cnn_filtered.vcf --offline --output_file /opt/vep/.vep/NSAIDS_0011.1d_cnn_filtered.txt</pre></td></tr><tr><td>Start time</td><td>2021-02-03 00:14:01</td></tr><tr><td>End time</td><td>2021-02-03 00:31:51</td></tr><tr><td>Run time</td><td>1070 seconds</td></tr><tr><td>Input file</td><td>/opt/vep/.vep/NSAIDS_0011.1d_cnn_filtered.vcf</td></tr><tr><td>Output file</td><td>/opt/vep/.vep/NSAIDS_0011.1d_cnn_filtered.txt</td></tr></table><h3 id="gen_stats">General statistics</h3><table class="stats_table"><tr><td>Lines of input read</td><td>377194</td></tr><tr><td>Variants processed</td><td>376686</td></tr><tr><td>Variants filtered out</td><td>0</td></tr><tr><td>Novel / existing variants</td><td>-</td></tr><tr><td>Overlapped genes</td><td>46102</td></tr><tr><td>Overlapped transcripts</td><td>202048</td></tr><tr><td>Overlapped regulatory features</td><td>-</td></tr></table><hr/><h3 id="var_class">Variant classes</h3><div id="var_class_pie" style="width: 800px; height: 200px">&nbsp;</div><div id="var_class_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="var_cons">Consequences (most severe)</h3><div id="var_cons_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="var_cons_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="consequences">Consequences (all)</h3><div id="consequences_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="consequences_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="coding">Coding consequences</h3><div id="coding_pie" style="width: 800px; height: 400px">&nbsp;</div><div id="coding_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="chr">Variants by chromosome</h3><div id="chr_bar" style="width: 800px; height: 200px">&nbsp;</div><div id="chr_table" style="width: 800px; height: 200px">&nbsp;</div><hr/><h3 id="protein">Position in protein</h3><div id="protein_bar" style="width: 800px; height: 200px">&nbsp;</div></div>
</div></body>
</html>
